File: PCSimulator/MDPSimulator/MDPModel/Robot.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.ComponentModel;
using System.Diagnostics;
namespace MDPModel
{
    public class Robot:INotifyPropertyChanged
    {
        public event PropertyChangedEventHandler PropertyChanged;
        private int x;
        private int y;
        public int X
        {   get{return x;}
            set
            {
                x = value;
                OnPropertyChanged("X");
            }
        }

        
        public int Y 
        { 
            get {return this.y;} 
            set
            {
                this.y = value;
                OnPropertyChanged("Y");
            } 
        }
        public int Dir{ get;set; }

        public int Range { get; set; }

        public Map Env { get; set; }

        public Map Memory { get; set; }

        public Robot(int x, int y, int r, char d)
        {
            this.X = x;
            this.Y = y;
            this.Range = r;
            this.Dir = d;
            //empty memory for robot
            this.Memory = new Map();
        }

        public Robot()
        {
            this.X = 1;
            this.Y = 1;
            this.Range = 1;
            this.Dir = 'U';
            this.Memory = new Map();
        }


        public void turnLeft()
        {
            switch (this.Dir)
            {
                case 'U':
                    this.Dir = 'L';
                    break;
                case 'D':
                    this.Dir = 'R';
                    break;
                case 'R':
                    this.Dir = 'U';
                    break;
                case 'L':
                    this.Dir = 'D';
                    break;
                default:
                    break;
            }
        }

        public void turnRight()
        {
            switch (this.Dir)
            {
                case 'U':
                    this.Dir = 'R';
                    break;
                case 'D':
                    this.Dir = 'L';
                    break;
                case 'R':
                    this.Dir = 'D';
                    break;
                case 'L':
                    this.Dir = 'U';
                    break;
                default:
                    break;
            }
        }


        public void moveForward(int dis)
        {
            switch (this.Dir)
            {
                case 'U':
                    this.Y+=dis;
                    break;
                case 'D':
                    this.Y-=dis;
                    break;
                case 'R':
                    this.X+=dis;
                    break;
                case 'L':
                    this.X-=dis;
                    break;
                default:
                    break;
            }
        }

        public void turnAround()
        {
            switch (this.Dir)
            {
                case 'U':
                    this.Dir = 'D';
                    break;
                case 'D':
                    this.Dir = 'U';
                    break;
                case 'R':
                    this.Dir = 'L';
                    break;
                case 'L':
                    this.Dir = 'R';
                    break;
                default:
                    break;
            }
        }

        public void simulateExplore()
        {
            bool isBlockedLeft, isBlockedRight, isBlockedFront;
            //turnRight();
            bool moved = false;
            do
            {
                Console.WriteLine("Current X = {0}, Y = {1}", this.X, this.Y);
                try
                {
                    isBlockedFront = scanFront();
                    isBlockedLeft = scanLeft();
                    isBlockedRight = scanRight();
                    if (!isBlockedRight)
                    {
                        Console.WriteLine("Turn right");
                        turnRight();
                        moveForward(1);
                    }
                    else if (!isBlockedFront)
                    {
                        Console.WriteLine("Move");
                        moved = true;
                        moveForward(1);
                    }
                    else if (!isBlockedLeft)
                    {
                        Console.WriteLine("Turn left");
                        turnLeft();
                        moveForward(1);
                    }
                    else
                    {
                        Console.WriteLine("Turn around");
                        turnAround();
                    }
                }
                catch (Exception e)
                {
                    Console.WriteLine(e.Message);
                }
            } while ((this.X != 1 || this.Y != 1) || !moved);
        }


        public bool scanFront()
        {
            switch (this.Dir)
            {
                case 'R':
                    return checkRightSide();
                case 'L':
                    return checkLeftSide();
                case 'U':
                    return checkUp();
                case 'D':
                    return checkDown();
                default:
                    return true;
            }    
        }

        public bool scanLeft()
        {
            switch (this.Dir)
            {
                case 'R':
                    return checkUp();
                case 'L':
                    return checkDown();
                case 'U':
                    return checkLeftSide();
                case 'D':
                    return checkRightSide();
                default:
                    return true;
            }
        }

        public bool scanRight()
        {
            switch (this.Dir)
            {
                case 'R':
                    return checkDown();
                case 'L':
                    return checkUp();
                case 'U':
                    return checkRightSide();
                case 'D':
                    return checkLeftSide();
                default:
                    return true;
            }
        }

        public bool checkUp()
        {
            bool isBlocked = false;
            if (this.Y + this.Range + 1 >= this.Env.Grid.GetLength(0))
                return true;
            int y = this.Y + this.Range + 1;
            for (int i = this.X - this.Range; i <= this.X + this.Range; i++)
            {
                if (this.Env.Grid[y, i].Status == 1)
                {
                    isBlocked = true;
                    //explored and has obstacle
                    this.Memory.Grid[y, i].Status = 1;
                }
                else
                {
                    //empty cell
                    this.Memory.Grid[y, i].Status = 2;
                }

            }
            return isBlocked;
        }

        
        public bool checkRightSide()
        {
            bool isBlocked = false;

            if (this.X + this.Range + 1 >= this.Env.Grid.GetLength(1))
                return true;
            int x = this.X + this.Range + 1;
            for (int i = this.Y - this.Range; i <= this.Y + this.Range; i++)
            {
                if (this.Env.Grid[i,x].Status == 1)
                {
                    isBlocked = true;
                    //explored and has obstacle
                    this.Memory.Grid[i, x].Status = 1;
                }
                else
                {
                    //empty cell
                    this.Memory.Grid[i, x].Status = 2;
                }
            }
            return isBlocked;
        }

        public bool checkLeftSide()
        {
            bool isBlocked = false;
            if (this.X - this.Range - 1 < 0)
                return true;
            int x = this.X - this.Range - 1;
            for (int i = this.Y - this.Range; i <= this.Y + this.Range; i++)
            {
                if (this.Env.Grid[i, x].Status == 1)
                {
                    isBlocked = true;
                    //explored and has obstacle
                    this.Memory.Grid[i, x].Status = 1;
                }
                else
                {
                    //empty cell
                    this.Memory.Grid[i, x].Status = 2;
                }
            }
            return isBlocked;
        }

        public bool checkDown()
        {
            bool isBlocked = false;
            if (this.Y - this.Range - 1 < 0)
                return true;
            int y = this.Y - this.Range - 1;
            for (int i = this.X - this.Range; i <= this.X + this.Range; i++)
            {
                if (this.Env.Grid[y, i].Status == 1)
                {
                    isBlocked = true;
                    //explored and has obstacle
                    this.Memory.Grid[y, i].Status = 1;
                }
                else
                {
                    //empty cell
                    this.Memory.Grid[y, i].Status = 2;
                }
            }
            return isBlocked;
        }

        private void OnPropertyChanged(string p)
        {
            PropertyChangedEventHandler handler = PropertyChanged;
            if (handler != null)
            {
                handler(this, new PropertyChangedEventArgs(p));
            }
        }
    }
}

File: Algorithm/MDPpy/src/Robot.py
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.

__author__ = "Yiko"
__date__ = "$2-Mar-2015 10:34:11 AM$"

from Map import *
from Node import *
from Cell import *

class Robot:
    
    def __init__(self):
       self.X = 1;
       self.Y = 1;
       self.Range = 1;
       self.Dir = 'U';
       "empty memory for robot"
       self.Memory = Map(15,20);
       self.XGoal = 13;
       self.YGoal = 18;
       self.XStart = 1;
       self.YStart = 1;
       self.StartNode = Node(self.XStart, self.YStart)
       self.GoalNode = Node(self.XGoal, self.YGoal)
       "self.ShortestPath = Node[]"
       self.VirtualMap = Map(15,20)
    
    def __init__(self, x, y, r, d):
        self.X = x
        self.Y = y
        self.Range = r
        self.Dir = d
        "empty memory for robot"
        self.Memory = Map(15,20)
        self.XGoal = 13
        self.YGoal = 18
        self.XStart = 1
        self.YStart = 1
        self.StartNode = Node(self.XStart, self.YStart)
        self.GoalNode = Node(self.XGoal, self.YGoal)
        "self.ShortestPath = Node[]"
        self.VirtualMap = Map(15,20)
        self.enteredGoal=True
    

   
    


    def turnLeft(self):
        if self.Dir=='U':
            self.Dir = 'L'
          
        elif self.Dir=='D':
            self.Dir = 'R'
            
        elif self.Dir== 'R':
            self.Dir = 'U'
            
        elif self.Dir== 'L':
            self.Dir = 'D'
            
 
                
    def turnRight(self):  
        if self.Dir== 'U':
            self.Dir = 'R'
            
        elif self.Dir== 'D':
            self.Dir = 'L'
            
        elif self.Dir== 'R':
            self.Dir = 'D'
            
        elif self.Dir== 'L':
            self.Dir = 'U'


    def moveForward(self,dis):
        if self.Dir== 'U':
            self.Y+=dis
            
        elif self.Dir== 'D':
            self.Y-=dis
            
        elif self.Dir== 'R':
            self.X+=dis
            
        elif self.Dir== 'L':
            self.X-=dis
            
 


    def turnAround(self):
        if self.Dir == 'U':
            self.Dir = 'D'
            
        elif self.Dir== 'D':
            self.Dir = 'U'
            
        elif self.Dir== 'R':
            self.Dir = 'L'
            
        elif self.Dir== 'L':
            self.Dir = 'R'
            
            

    def explore(self,ArStr):
        moved = False
        while  ((self.X != 1 or self.Y != 1) or (not moved)):
            print "Current X = {0}, Y = {1}", self.X, self.Y
            try:
                isBlockedLeft = self.checkLeftSide(ArStr)
                isBlockedFront = self.checkTopSide(ArStr)
                isBlockedRight = self.checkRightSide(ArStr)
                if (not isBlockedRight):
                    print("Turn right")
                    self.turnRight()
                    return '2'
                elif (not isBlockedFront):
                    print("Move Forward");
                    moved = True
                    self.moveForward(1)
                    return '1'
                elif (not isBlockedLeft):
                    print("turn left");
                    self.turnLeft();
                    return '3'
                else:
                    print("Turn around")
                    self.turnAround()
                    return '4'
            except ValueError as e:
                print(e.Message)
                print(e)
    
    def checkLeftSide(self, ArStr):
        # left here
        isBlocked = False
  

        #x = self.X - self.Range - 1
        if (ArStr[0] == '1'):
            isBlocked = True
            "explored and has obstacle"
            #self.Memory.grid[self.Y][x].Status = 1
        elif (ArStr[0] == '2'):
            "map second block to have obstacle"
            #self.Memory.grid[][]
        else:
            "empty cell"
            #self.Memory.grid[self.Y][x].Status = 2
        print "Left", isBlocked
        return isBlocked
    
    def checkTopSide(self, ArStr):
        isBlocked = False
        #if (self.Y + self.Range + 1 >= self.Env.Grid.GetLength(0))
        #    return True;
        # y = self.Y + self.Range + 1
     

        for i in range(1,4):
            if (ArStr[i]=='1'):
                isBlocked = True
                "explored and has obstacle"
                #self.Memory.grid[y][i].Status = 1
            else:
                "empty cell"
                #self.Memory.grid[y][i].Status = 2
        print "TOP", isBlocked
        return isBlocked


    def checkRightSide(self, ArStr):
        isBlocked = False
       # if (this.X + this.Range + 1 >= this.Env.Grid.GetLength(1))
        #    return True;
        #x = self.X + self.Range + 1

        for i in range(4,6):
            if (ArStr[i] == '1'):
                isBlocked = True
                "explored and has obstacle"
                #self.Memory.grid[self.Y][x].Status = 1
            else:
                "empty cell"
                #self.Memory.grid[self.Y][x].Status = 2
        print "RIGHT", isBlocked
        return isBlocked


    def generateMapStr(self):
        # first 5 characters: direction + current center position
        if (self.X < 10):
            x = '0'+ str(self.X)
        else:
            x = str(self.X)
        if (self.Y < 10):
            y = '0'+ str(self.Y)
        else:
            y = str(self.Y) 
        mapStr = self.Dir + x + y
        
        # iterate the map to add value to mapStr
        for i in range (20):
            for j in range (15):
                mapStr += str(self.Memory.grid[i][j].status)
        if (self.X == 1 and self.Y == 1 and self.enteredGoal):
            mapStr += 'F'
        return mapStr